prow/plugins/dog/dog_test.go

```diff
@@ -0,0 +1,229 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dog
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"regexp"
+	"strings"
+	"testing"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/github/fakegithub"
+)
+
+type fakePack string
+
+var human = flag.Bool("human", false, "Enable to run additional manual tests")
+
+func (c fakePack) readDog() (string, error) {
+	return string(c), nil
+}
+
+func TestRealDog(t *testing.T) {
+	if !*human {
+		t.Skip("Real dogs disabled for automation. Manual users can add --human [--category=foo]")
+	}
+	if cat, err := dogURL.readDog(); err != nil {
+		t.Errorf("Could not read dog from %s: %v", dogURL, err)
+	} else {
+		fmt.Println(cat)
+	}
+}
+
+func TestFormat(t *testing.T) {
+	re := regexp.MustCompile(`\[!\[.+\]\(.+\)\]\(.+\)`)
+	basicURL := "http://example.com"
+	testcases := []struct {
+		name string
+		url  string
+		err  bool
+	}{
+		{
+			name: "basically works",
+			url:  basicURL,
+			err:  false,
+		},
+		{
+			name: "empty url",
+			url:  "",
+			err:  true,
+		},
+		{
+			name: "bad url",
+			url:  "http://this is not a url",
+			err:  true,
+		},
+	}
+	for _, tc := range testcases {
+		ret, err := dogResult{
+			URL: tc.url,
+		}.Format()
+		switch {
+		case tc.err:
+			if err == nil {
+				t.Errorf("%s: failed to raise an error", tc.name)
+			}
+		case err != nil:
+			t.Errorf("%s: unexpected error: %v", tc.name, err)
+		case !re.MatchString(ret):
+			t.Errorf("%s: bad return value: %s", tc.name, ret)
+		}
+	}
+}
+
+// Medium integration test (depends on ability to open a TCP port)
+func TestHttpResponse(t *testing.T) {
```

Comment:
Seems like this wants to have a testcase struct with []byte response, url string, err book } or something like that and then cases such as "good" "http error" "not json" "wrong keys" "not a url"

Reply:
SGTM, pretty much did just this.